<a href="https://colab.research.google.com/github/DevTedd/IndividualProjectMoringaWeek3/blob/main/Moringa_Data_Science_Prep_W3_Independent_Project_2021_09_Ted_Kimani_DataReport.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Friday: Independent Project - Week 3

In this week's independent project, you will be working as Data Scientist for MTN Cote d'Ivoire, a leading telecom company and you will be solving for the following research question.

Currently MTN Cote d'Ivoire would like to upgrade its technology infrastructure for its mobile users in Ivory Coast. Studying the given dataset, how does MTN Cote d'Ivoire go about the upgrade of its infrastructure strategy within the given cities?

Your final deliverable will be a Data Report which will comprise the following sections;




1.	Business Understanding 
2.	Data Understanding 
3.	Data Preparation 
4.	Analysis 
5.	Recommendation 
6.	Evaluation


You can use the CRISP-DM methodology to guide you while working on the Data Report. Your Data Report will also need to have an objective account, with insights majorly coming from the dataset.

# Importing libraries and datasets

In [2]:
import pandas as pd
# as well as the Numpy library
import numpy as np

In [3]:
telcomDf1 = pd.read_csv('/content/drive/MyDrive/Individual Project Week 3/Telcom_dataset.csv')
telcomDf1.head(5)
telcomDf1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5001 entries, 0 to 5000
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   PRODUTC          5001 non-null   object
 1   VALUE            5001 non-null   int64 
 2   DATETIME         5001 non-null   object
 3   CELL_ON_SITE     5001 non-null   object
 4   DW_A_NUMBER_INT  5001 non-null   object
 5   DW_B_NUMBER_INT  5001 non-null   object
 6   COUNTRY_A        5001 non-null   object
 7   COUNTRY_B        5001 non-null   object
 8   CELL_ID          5001 non-null   object
 9   SITE_ID          4285 non-null   object
dtypes: int64(1), object(9)
memory usage: 390.8+ KB


In [4]:
telcomDf2 = pd.read_csv('/content/drive/MyDrive/Individual Project Week 3/Telcom_dataset2.csv')
telcomDf2.head(5)
#telcomDf2.info()

PRODUCT  VALUE              DATE_TIME  ... COUNTRY_B     CELL_ID     SITE_ID
0     sms      0  2012-05-07 23:02:06.0  ...     19e2e  +854AcBQT2  +854AcBQT2
1   Voice      0  2012-05-07 23:03:44.0  ...     19e2e  +854AcBQT2  +854AcBQT2
2     sms      0  2012-05-07 23:04:06.0  ...     19e2e  +laSrk7g6q  +laSrk7g6q
3   Voice     37  2012-05-07 23:04:48.0  ...     19e2e  /xvRrCVKoQ  /xvRrCVKoQ
4     sms     25  2012-05-07 23:04:58.0  ...     19e2e  +bO+qkSonO  +bO+qkSonO

[5 rows x 10 columns]

In [67]:
telcomDf3 = pd.read_csv('/content/drive/MyDrive/Individual Project Week 3/Telcom_dataset3.csv')
telcomDf3.shape

(5001, 10)

In [66]:
cellsGeo = pd.read_csv('/content/drive/MyDrive/Individual Project Week 3/cells_geo.csv', sep=';')
cellsGeo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3974 entries, 0 to 3973
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Unnamed: 0    3974 non-null   int64  
 1   VILLES        3974 non-null   object 
 2   STATUS        3907 non-null   object 
 3   LOCALISATION  3974 non-null   object 
 4   DECOUPZONE    3974 non-null   object 
 5   ZONENAME      3968 non-null   object 
 6   LONGITUDE     3974 non-null   float64
 7   LATITUDE      3974 non-null   float64
 8   REGION        3974 non-null   object 
 9   AREA          3951 non-null   object 
 10  CELL_ID       3974 non-null   object 
 11  SITE_CODE     3974 non-null   object 
dtypes: float64(2), int64(1), object(9)
memory usage: 372.7+ KB


In [ ]:
#Theres 2 types of localization, within Abidjan and the Interiror 
B = cellsGeo.LOCALISATION.unique()
B


array(['ABIDJAN', 'INTERIEUR'], dtype=object)

In [7]:
cdrDescrip = pd.read_excel('/content/drive/MyDrive/Individual Project Week 3/CDR_description.xlsx')
cdrDescrip


Column name                                        Description   Format
0          PRODUCT                                       Voice or SMS   String
1            VALUE                                      Billing price  Integer
2        DATE_TIME               Time in format yyyy-MM-dd hh:mm:ss.0   String
3     CELL_ON_SITE  Which cell in the site was used (not needed here)  Integer
4  DW_A_NUMBER_INT  Anonymized phone number of the person for whic...   String
5  DW_B_NUMBER_INT        Anonymized phone number of the counterparty   String
6        COUNTRY_A                  Country of party A (useless here)   String
7        COUNTRY_B                  Country of party B (useless here)   String
8          CELL_ID                                     ID of the cell   String
9          SITE_ID                                     ID of the SITE   String

In [8]:
cellsDescrip = pd.read_excel('/content/drive/MyDrive/Individual Project Week 3/cells_geo_description.xlsx')
cellsDescrip

Column name                                        Description  Format
0         VILLES                                               City  String
1         STATUS                                  In Service or not  String
2   LOCALISATION                                  in ABIDJAN or not  String
3     DECOUPZONE                                  Geographical Zone  String
4       ZONENAME                                       Name of Zone  String
5      LONGITUDE                                          Longitude   Float
6       LATITUDE                                           Latitude   Float
7         REGION                                             Region  String
8           AREA                                               Area  String
9        CELL_ID                                     ID of the cell  String
10     SITE_CODE  Site (there are several cells per site, severa...  String

We will begin by finding out the number of cities in out data set cellsGeo

In [9]:
#This shows us there is around 504 cities, with is quite large
cities = cellsGeo['VILLES'].drop_duplicates()
cities.count()


504

In [ ]:
#We know look for the number of unique geographical zones, which are alot less compared to number of cites
geoZones = cellsGeo['DECOUPZONE'].drop_duplicates()
geoZones



0          "Abidjan_EST
3             Grand-EST
6           Grand-OUEST
7         "Abidjan_NORD
18      "Abidjan_CENTRE
23         Abidjan_NORD
35          Abidjan_EST
122       Abidjan_OUEST
180           Grand-SUD
216           SUD-OUEST
249      Abidjan_CENTRE
271         Abidjan_SUD
372          Grand-NORD
788          Gd-ABIDJAN
1559                  0
Name: DECOUPZONE, dtype: object

In [ ]:
#We will group the data into Regions and cities. The regions are a good place to start 
B = cellsGeo.REGION.unique()
B
geoZonesGroup = cellsGeo.groupby(['VILLES','REGION'])
geoZonesGroup.first()
#Not much can be gained from this

Unnamed: 0      STATUS  ...     CELL_ID   SITE_CODE
VILLES        REGION                                  ...                        
ABENGOUROU    MOYEN COMOE             44  In Service  ...  8cdfc1bc65  6e87d6cb59
ABOBO         LAGUNE                   7  In Service  ...  a70c1d263b  56306cc7ca
ABOISSO       SUD COMOE              158  In Service  ...  b4d8c16d9e  38c5c1de07
ABOISSO COMOE LAGUNE                 155  In Service  ...  ff4372eafb  640427c4c9
ABONGOUA      MOYEN COMOE            177  In Service  ...  d1a1270467  a8d8462dda
...                                  ...         ...  ...         ...         ...
ZATTA         LACS                  3941  In Service  ...  1d41c565fc  c644d1ce4c
ZIKISSO       SUD BANDAMA           3944  In Service  ...  733df73327  211015b14d
ZON-HOUNIEN   18 MONTAGNES          3960  In Service  ...  a3e5534191  32b8357a82
ZOUKOUGBEU    HAUT SASSANDRA        3963  In Service  ...  317ba2e3e8  443ac3fc19
ZUENOULA      MARAHOUE              3382  In Service  ...  8cde542067  43adae9a57

[508 rows x 10 columns]

In [10]:
localZones = cellsGeo.groupby(['LOCALISATION'])
localZones.count()

Unnamed: 0  VILLES  STATUS  ...  AREA  CELL_ID  SITE_CODE
LOCALISATION                              ...                          
ABIDJAN             1669    1669    1636  ...  1650     1669       1669
INTERIEUR           2305    2305    2271  ...  2301     2305       2305

[2 rows x 11 columns]

From this we can see that the netork can be divided into 2 localizatios, with most cities and cellphones at 2305 are outside Abidjan with 1669 left inside

# We will now merge the three days data with the geographical data

In [32]:
#This is the first days telcom data
dayOne_A = telcomDf1
dayOne_A = dayOne_A.drop_duplicates()



dayOne_Loc = cellsGeo
dayOne_Loc = dayOne_Loc.drop_duplicates()
dayOne_combines = pd.merge(dayOne_A,dayOne_Loc, left_on='SITE_ID', right_on='SITE_CODE')

dayOne_combinesValue = dayOne_combines['VALUE'].sum()

dayOne_combines#the entire combined data set
#dayOne_Av#The one wthout rows of no value
#dayOne_combines['VALUE'].median()
#dayOne_combines['VALUE'].var()
#dayOne_combines .corr()


PRODUTC  VALUE  ...   CELL_ID_y   SITE_CODE
0         sms      0  ...  24ccdda3d5  1b5540c02d
1         sms      0  ...  ebfc58af70  1b5540c02d
2         sms      0  ...  ba8ce6e4ef  1b5540c02d
3         sms      0  ...  c2545171dc  1b5540c02d
4         sms      0  ...  f15bd5483f  1b5540c02d
...       ...    ...  ...         ...         ...
16274   Voice      0  ...  e899016a7d  f72cd95f00
16275   Voice      0  ...  f4271c1bc1  f72cd95f00
16276   Voice      1  ...  1d41c565fc  c644d1ce4c
16277   Voice      1  ...  eb42ca8859  c644d1ce4c
16278   Voice      1  ...  7d655658f0  c644d1ce4c

[16279 rows x 22 columns]

In [27]:
dayTwo_A = telcomDf1
dayTwo_A = dayOne_A.drop_duplicates()
dayTwo_Loc = cellsGeo
dayTwo_loc = dayTwo_Loc.drop_duplicates()
dayTwo_combines = pd.merge(dayTwo_A,dayTwo_Loc, left_on='SITE_ID', right_on='SITE_CODE')


dayTwo_combinesValue = dayTwo_combines['VALUE'].sum()
dayTwo_combinesValue


475423

In [26]:
dayThree_A = telcomDf1
dayThree_A = dayThree_A.drop_duplicates()
dayThree_B = cellsGeo
dayThree_B = dayThree_B.drop_duplicates()
dayThree_combines = pd.merge(dayThree_A,dayThree_B, left_on='SITE_ID', right_on='SITE_CODE')

dayThree_combinesValue = dayThree_combines['VALUE'].sum()
dayThree_combinesValue

475423

In [33]:
print("The total value from communications on day one is: ",dayOne_combinesValue," And the second day's total is : ",dayTwo_combinesValue, "  and the third day's total value is: ",dayThree_combinesValue)

The total value from communications on day one is:  475423  And the second day's total is :  475423   and the third day's total value is:  475423


I will begin by dropping the columns that increase the noise

In [53]:
#This list shows how many individual phones are in a particular city for day 1
dayOne = dayOne_combines[['VILLES','CELL_ID_x','STATUS','LOCALISATION','DECOUPZONE','ZONENAME','REGION','AREA']]
groupCityDay1 = dayOne['CELL_ID_x'].groupby(dayOne['VILLES'])
groupCityDay1.size()

VILLES
ABENGOUROU            69
ABOBO               1993
ABOISSO               24
ABOISSO COMOE          6
ABOUDEKOUASSIKRO      15
                    ... 
ZATTA                  3
ZIKISSO               12
ZON-HOUNIEN            6
ZOUKOUGBEU             6
ZUENOULA              69
Name: CELL_ID_x, Length: 306, dtype: int64

In [54]:
#This list shows how many individual phones are in a particular city for day 2
dayTwo = dayTwo_combines[['VILLES','CELL_ID_x','STATUS','LOCALISATION','DECOUPZONE','ZONENAME','REGION','AREA']]
groupCityDay2 = dayTwo['CELL_ID_x'].groupby(dayTwo['VILLES'])
groupCityDay2.size()

VILLES
ABENGOUROU            69
ABOBO               1993
ABOISSO               24
ABOISSO COMOE          6
ABOUDEKOUASSIKRO      15
                    ... 
ZATTA                  3
ZIKISSO               12
ZON-HOUNIEN            6
ZOUKOUGBEU             6
ZUENOULA              69
Name: CELL_ID_x, Length: 306, dtype: int64

In [52]:
#This list shows how many individual phones are in a particular city for day 3
dayThree = dayThree_combines[['VILLES','CELL_ID_x','STATUS','LOCALISATION','DECOUPZONE','ZONENAME','REGION','AREA']]
groupCityDay3 = dayThree['CELL_ID_x'].groupby(dayThree['VILLES'])
groupCityDay3.size()

VILLES
ABENGOUROU            69
ABOBO               1993
ABOISSO               24
ABOISSO COMOE          6
ABOUDEKOUASSIKRO      15
                    ... 
ZATTA                  3
ZIKISSO               12
ZON-HOUNIEN            6
ZOUKOUGBEU             6
ZUENOULA              69
Name: CELL_ID_x, Length: 306, dtype: int64

The next steps is to order the three day dataset into descending order of cell phone density per city. Thats a priority list of which have the most users and should be focused first. Upgrading denser areas allows us to generate more revnue from these areas and use the incremetal funds to upgrade the less desnse are thenafer. If the days are similiar in all 3 days those are consistently used areas


In [59]:
#adding the size of 
#dayOne.sort_values(['CELL_ID_x','VILLES'], ascending= False).groupby('VILLES').first()
dayOne['VILLES'].describe()                                                                                  

count        16279
unique         306
top       YOPOUGON
freq          2686
Name: VILLES, dtype: object

In [62]:
#Ordered list of day 1 based on phone population density per city
groupCityDay1.size().sort_values(ascending = False)

VILLES
YOPOUGON        2686
ABOBO           1993
COCODY          1062
YAMOUSSOUKRO     831
ADJAME           797
                ... 
LATAHA             3
DJENIDOUFLA        3
DJOUKOUYO          3
SAMATIGUILA        3
KOTOULA            2
Name: CELL_ID_x, Length: 306, dtype: int64

In [61]:
#Ordered list of day 2 based on phone population density per city
groupCityDay2.size().sort_values(ascending = False)

VILLES
YOPOUGON        2686
ABOBO           1993
COCODY          1062
YAMOUSSOUKRO     831
ADJAME           797
                ... 
LATAHA             3
DJENIDOUFLA        3
DJOUKOUYO          3
SAMATIGUILA        3
KOTOULA            2
Name: CELL_ID_x, Length: 306, dtype: int64

In [63]:
#Ordered list of day 3 based on phone population density per city
groupCityDay3.size().sort_values(ascending = False)

VILLES
YOPOUGON        2686
ABOBO           1993
COCODY          1062
YAMOUSSOUKRO     831
ADJAME           797
                ... 
LATAHA             3
DJENIDOUFLA        3
DJOUKOUYO          3
SAMATIGUILA        3
KOTOULA            2
Name: CELL_ID_x, Length: 306, dtype: int64